### PR TITLE
Ansible 4

### PR DESCRIPTION
**Pull request by kumite73**
# Выполнено ДЗ Ansible 4

 - [X] Основное ДЗ
 - [X] Задание со *

## В процессе сделано:
Устанавливаем `virtualbox`
Устанавливаем `vagrant 2.0.4` 
```
wget https://releases.hashicorp.com/vagrant/2.0.4/vagrant_2.0.4_x86_64.deb
sudo dpkg -i vagrant_2.0.4_x86_64.deb
```
Так-как я не смог добиться, чтобы 2.1.2 версия не падала с ошибкой

    vagrant provision dbserver

    ==> dbserver: Running provisioner: ansible...
    dbserver: Running ansible-playbook...
    ImportError: No module named site
    Ansible failed to complete successfully. Any error output should be
    visible above. Please fix these errors and try again.

Добавляем следующие строки в `.gitignore`

    # Vagrant & molecule
    .vagrant/
    *.log
    *.pyc
    .molecule
    .cache
    .pytest_cache

Создаем файл `Vagrantfile`

    Vagrant.configure("2") do |config|

      config.vm.provider :virtualbox do |v|
        v.memory = 512
      end

      config.vm.define "dbserver" do |db|
        db.vm.box = "ubuntu/xenial64"
        db.vm.hostname = "dbserver"
        db.vm.network :private_network, ip: "10.10.10.10"
      end

      config.vm.define "appserver" do |app|
        app.vm.box = "ubuntu/xenial64"
        app.vm.hostname = "appserver"
        app.vm.network :private_network, ip: "10.10.10.20"
      end
    end

Выполняем `vagrant up`
Смотрим скачанные образы `vagrant box list`
Проверям статус VMs `vagrant status`
Проверим SSH доступ к VM `vagrant ssh appserver`
Добавим провижинер

    db.vm.provision "ansible" do |ansible|
      ansible.playbook = "playbooks/site.yml"
      ansible.groups = {
      "db" => ["dbserver"],
      "db:vars" => {"mongo_bind_ip" => "0.0.0.0"}
      }
    end

Запуск провижинера `vagrant provision dbserver`
Создаем `base.yml` для установки питона

    ---
    - name: Check && install python
      hosts: all
      become: true
      gather_facts: False

      tasks:
        - name: Install python for Ansible
          raw: test -e /usr/bin/python || (apt -y update && apt install -y python-minimal)
          changed_when: False

Меняем `site.yml`

    ---
    - include: base.yml
    - include: db.yml
    - include: app.yml
    - include: deploy.yml

Проверяем `vagrant provision dbserver`
Изменим роль `db`, добавив файл тасков `db/tasks/install_mongo.yml`

    - name: Add APT key
      apt_key:
        id: "EA312927"
        keyserver: keyserver.ubuntu.com
      tags: install

    - name: Add APT repository
      apt_repository:
        repo: deb [ arch=amd64,arm64 ] http://repo.mongodb.org/apt/ubuntu xenial/mongodb-org/3.2 multiverse
        state: present
      tags: install

    - name: Install mongodb package
      apt:
        name: mongodb-org
        state: present
      tags: install

    - name: Configure service supervisor
      systemd:
        name: mongod
        enabled: yes
        state: started
      tags: install

Создаем `db/tasks/config_mongo.yml`

    ---
    - name: Change mongo config file
      template:
        src: templates/mongod.conf.j2
        dest: /etc/mongod.conf
        mode: 0644
      notify: restart mongod

Меняем `db/tasks/main.yml`

    ---
    # tasks file for db

    - name: Show info about the env this host belongs to
      debug:
        msg: "This host is in {{ env }} environment!!!"

    - include: install_mongo.yml
    - include: config_mongo.yml

Проверяем `vagrant provision dbserver`
Создаем `app/tasks/ruby.yml`

    ---
    - name: Install ruby and rubygems and required packages
      apt: "name={{ item }} state=present"
        with_items:
          - ruby-full
          - ruby-bundler
          - build-essential
        tags: ruby

Создаем `app/tasks/puma.yml`

    ---
    - name: Add unit file for Puma
      copy:
        src: puma.service
        dest: /etc/systemd/system/puma.service
      notify: reload puma

    - name: Add config for DB connection
      template:
        src: db_config.j2
        dest: /home/appuser/db_config
        owner: appuser
        group: appuser

      - name: enable puma
        systemd: name=puma enabled=yes

Меняем `app/tasks/main.yml`

    ---
    # tasks file for app

    - name: Show info about the env this host belongs to
      debug:
        msg: "This host is in {{ env }} environment!!!"

    - include: ruby.yml
    - include: puma.yml

Провижиним `appserver`

    app.vm.provision "ansible" do |ansible|
      ansible.playbook = "playbooks/site.yml"
      ansible.groups = {
      "app" => ["appserver"],
      "app:vars" => { "db_host" => "10.10.10.10"}
      }
    end

Проверяем `vagrant provision appserver`

#### параметризуем модули

Меняем `puma.yml`

    ---
    - name: Add unit file for Puma
      template:
        src: puma.service.j2
        dest: /etc/systemd/system/puma.service
      notify: reload puma

    - name: Add config for DB connection
      template:
        src: db_config.j2
        dest: "/home/{{ deploy_user }}/db_config"
        owner: "{{ deploy_user }}"
        group: "{{ deploy_user }}"

    - name: enable puma
      systemd: name=puma enabled=yes

Перенносим его в `app/tasks/puma.yml`
Меняем `ansible/playbooks/deploy.yml`

    - name: Deploy App
      hosts: app
      vars:
        deploy_user: appuser

      tasks:
        - name: Fetch the latest version of application code
          git:
            repo: 'https://github.com/express42/reddit.git'
            dest: "/home/{{ deploy_user }}/reddit"
            version: monolith
          notify: restart puma

        - name: bundle install
          bundler:
            state: present
            chdir: "/home/{{ deploy_user }}/reddit"

      handlers:
        - name: restart puma
          become: true
          systemd: name=puma state=restarted

Переопределяем переменные провиженера

    ansible.extra_vars = {
      "deploy_user" => "vargant"
    }

Проверка `vagrant provision appserver`

### Задание со *

Добавляем переменные в файл `/ansible/roles/app/vars/mailn.yml`

    nginx_sites:
      default:
        - listen {{ jdauphant_nginx_listen_port }}
        - server_name {{ jdauphant_nginx_server_name }}
        - |
          location / {
            proxy_pass http://127.0.0.1:{{ jdauphant_nginx_proxy_pass_port }};
          }

Выполняем `vagrant provision appserver`

### Тестирование роли

Меняем `requirements.txt`

    ansible>=2.4
    molecule>=2.6
    testinfra>=1.10
    python-vagrant>=0.5.15

Выполняем установку `pip install -r requirements.txt`

Выполняем команду в директории с ролью `ansible/roles/db`

    molecule init scenario --scenario-name default -r db -d vagrant

Создаем `db/molecule/default/tests/test_default.py`

    import os

    import testinfra.utils.ansible_runner

    testinfra_hosts = testinfra.utils.ansible_runner.AnsibleRunner(
        os.environ['MOLECULE_INVENTORY_FILE']).get_hosts('all')

    # check if MongoDB is enabled and running
    def test_mongo_running_and_enabled(host):
        mongo = host.service("mongod")
        assert mongo.is_running
        assert mongo.is_enabled

    # check if configuration file contains the required line
    def test_config_file(host):
        config_file = host.file('/etc/mongod.conf')
        assert config_file.contains('bindIp: 0.0.0.0')
        assert config_file.is_file

Запускаем в директории `ansible/roles/db`

    molecule create

Подключиться по SSH

    molecule login -h instance

Изменяем `db/molecule/default/playbook.yml`

    ---
    - name: Converge
      become: true
      hosts: all
      vars:
        mongo_bind_ip: 0.0.0.0
      roles:
        - role: db

Применим `playbook.yml`

    molecule converge

Запускаем тесты `molecule verify`

Добавим в файл новый тест `db/molecule/default/tests/test_default.py`

    # check if MongoDB is listening port
    def test_mongo_listen_port(host):
        assert host.socket("tcp://0.0.0.0:27017").is_listening

Запускаем тесты `molecule verify`

Меняем плейбук `playbooks/packer_db.yml`

    ---
    - name: Install MongoDB 3.2
      hosts: all
      become: true
      roles:
        - db

Запускаем билд `packer build -var-file=packer/variables.json packer/db.json`
Не видит роли ansible
Меняем `packer/packer_db.yml`

    "provisioners": [{
      "type": "ansible",
      "playbook_file": "ansible/playbooks/packer_db.yml",
      "ansible_env_vars": ["ANSIBLE_ROLES_PATH={{ pwd }}/ansible/roles"]   # Добавляем пути к ролям
    }]

Запускаем билд `packer build -var-file=packer/variables.json packer/db.json`

Меняем плейбук `playbooks/packer_app.yml`

    ---
    - name: Install Ruby && Bundler
      hosts: all
      become: true
      roles:
        - app

Меняем `packer/packer_app.yml`

    "provisioners": [{
      "type": "ansible",
      "playbook_file": "ansible/playbooks/packer_app.yml",
      "ansible_env_vars": ["ANSIBLE_ROLES_PATH={{ pwd }}/ansible/roles"]   # Добавляем пути к ролям
    }]

Запускаем билд `packer build -var-file=packer/variables.json packer/app.json`

## PR checklist
 - [X] Выставил label с номером домашнего задания
 - [X] Выставил label с темой домашнего задания

### CODE REVIEW

**Review by whitew1nd — APPROVED** [review]
Looks good for me, что касается основного задания.
По заданию со звездочкой есть нюанс: в описании задания предлагалось перенести конфигурацию проксирования в Vagrantfile, а не в переменные роли

File: ansible/Vagrantfile [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    app.vm.box = "ubuntu/xenial64"
+    app.vm.hostname = "appserver"
+    app.vm.network :private_network, ip: "10.10.10.20"
+
+    app.vm.provision "ansible" do |ansible|
+      #ansible.compatibility_mode = "1.8"
+      ansible.playbook = "playbooks/site.yml"
+      ansible.groups = {
+        "app" => ["appserver"],
+        "app:vars" => {"db_host" => "10D.10.10.10"}
```
Comment (whitew1nd) [inline ansible/Vagrantfile:32]: а почему получился такой адрес? `"10D.10.10.10"`

Comment (kumite73) [reply ansible/Vagrantfile:32]: Я работаю на винде сервер на убунте. Все шорткаты разные. Сначала я все запускаю и тестирую, потом уже пишу README.   100% опечатка. Исправлю

Comment (kumite73) [reply ansible/Vagrantfile:32]: По vagrant Я не смог сформировать валидный хеш для этой конструкции.

Comment (kumite73) [reply ansible/Vagrantfile:32]: # vars file for app
nginx_sites:
  default:
    - listen 80
    - server_name reddit
    - |
      location / {
        proxy_pass http://127.0.0.1:9292;
      }

Перепробовал массу вариантов. Не проходт. Я с руби 10 лет, но тут видимо непонимание что хочет внешняя роль.. Если ткнете носом буду очень признателен!!!

Comment (whitew1nd) [reply ansible/Vagrantfile:32]: Не вопрос, держи подсказку:
В переменной nginx_sites у нас ассоциативный массив с одним ключом default, значением которого является массив, где каждый элемент – строка с директивами конфигурации nginx
```
"nginx_sites" => {
  "default": [ "listen 80", ...]
}
```

**Comment by whitew1nd** [comment]
В остальном, можно мерджить

**Pull request merged**